sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1753,15 +1753,26 @@ class DataFrame private[sql](
    * Converts a JavaRDD to a PythonRDD.
    */
   protected[sql] def javaToPython: JavaRDD[Array[Byte]] = {
-    val structType = schema  // capture it for closure
-    val rdd = queryExecution.toRdd.map(EvaluatePython.toJava(_, structType))
-    EvaluatePython.javaToPython(rdd)
+    if (isOutputPickled) {
+      queryExecution.toRdd.map(_.getBinary(0))
```

Comment:
I introduced this for aggregate at first.
Think about `ds.map(...).groupByKey(key_func, key_schema).mapGroups(...)`, when do aggregate, we need to run the key function and append key data to the input rows. In this case, the input rows are the python data after `map`. So we need to store the python data at JVM side even without schema, but the row count should be corrected.

It's also more clear when we call `ds.map(...).count()`, users would expect same row count after typed operations.